File: leetcode/74. Search a 2D Matrix/sol.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int lRow = 0, rRow = matrix.size()-1;
        int lCol = 0, rCol = matrix[0].size()-1;
        while(lRow < rRow){
            int midRow = lRow + (rRow- lRow)/2;
            if(matrix[midRow][0] == target) return true;
            else if(matrix[midRow][0] < target && target < matrix[midRow+1][0]) {
                lRow = midRow;
                break;
            }
            else if(matrix[midRow][0] < target) lRow = midRow + 1;
            else rRow = midRow-1;
        }
        int row = lRow;
        while(lCol <= rCol){
            int midCol = lCol + (rCol- lCol)/2;
            if(matrix[row][midCol] == target) return true;
            else if(matrix[row][midCol] < target) lCol = midCol+1;
            else rCol = midCol-1;
        }
        return false;
    }
};
int main(){
    Solution s;
    vector<vector<int>> matrix = {{1}, {3}};
    int target = 3;
    bool ans = s.searchMatrix(matrix, target);
    cout << ans << endl;
    return 0;
}

File: 74. Search a 2D Matrix/sol.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        top, bot = 0, len(matrix) -1
        left, right = 0, len(matrix[0]) -1
        while top <= bot:   
            row = top + (bot - top) // 2
            if target > matrix[row][-1]:
                top = row + 1
            elif target < matrix[row][0]:
                bot = row - 1
            else:
                break

        if not(top <=bot):
            return False
            
        while left <= right:
            mid = left + (right - left) // 2
            if target > matrix[row][mid]:
                left = mid + 1
            elif target < matrix[row][mid]:
                right = mid - 1
            else:
                return True
        return False
